# Atividade Prática — QA com APIs (SWAPI)

## Versão do aluno — com scripts base

**Tempo estimado:** 60 minutos  
**Ferramenta:** Google Colab  
**Formato:** em grupos de **2 a 4 componentes** ou, se preferirem, **individualmente**.  
**Objetivo:** investigar respostas de uma API, ajustar scripts simples e responder às perguntas como um QA.

### Instruções
Nesta atividade, vocês **não precisam criar tudo do zero**.  
Os scripts já estão prontos como base. A tarefa do grupo será:

1. **Executar** os blocos de código
2. **Alterar apenas o que for pedido**
3. **Observar os resultados**
4. **Responder às perguntas em Markdown**
5. **Discutir em grupo** quais validações um QA faria

### Entrega
Cada grupo pode entregar **um único notebook** com:
- nomes dos integrantes
- respostas das perguntas
- scripts ajustados
- asserts completados

Se alguém optar por fazer sozinho, pode entregar individualmente sem problema.


In [4]:
import requests
import json

# -------------------------
# MODO DE SEGURANÇA (fallback)
# -------------------------
# Se a API não responder, o notebook usa dados locais
# para a atividade continuar funcionando.

DADOS_FALLBACK = {
    "https://swapi.dev/api/people/10/": {
        "name": "Obi-Wan Kenobi",
        "height": "182",
        "mass": "77",
        "eye_color": "blue-gray",
        "gender": "male",
        "url": "https://swapi.dev/api/people/10/"
    },
    "https://swapi.dev/api/people/13/": {
        "name": "Chewbacca",
        "height": "228",
        "mass": "112",
        "eye_color": "blue",
        "gender": "male",
        "url": "https://swapi.dev/api/people/13/"
    },
    "https://swapi.dev/api/planets/2/": {
        "name": "Alderaan",
        "climate": "temperate",
        "terrain": "grasslands, mountains",
        "population": "2000000000",
        "url": "https://swapi.dev/api/planets/2/"
    },
    "https://swapi.dev/api/starships/10/": {
        "name": "Millennium Falcon",
        "model": "YT-1300 light freighter",
        "manufacturer": "Corellian Engineering Corporation",
        "max_atmosphering_speed": "1050",
        "url": "https://swapi.dev/api/starships/10/"
    }
}

class RespostaLocal:
    def __init__(self, status_code, data):
        self.status_code = status_code
        self._data = data
        self.text = json.dumps(data, ensure_ascii=False)

    def json(self):
        return self._data

def buscar_endpoint(url):
    try:
        resposta = requests.get(url, timeout=10)
        if resposta.status_code == 200:
            return resposta, "online"
    except Exception:
        pass

    if url in DADOS_FALLBACK:
        return RespostaLocal(200, DADOS_FALLBACK[url]), "fallback local"

    return RespostaLocal(404, {"detail": "Not found"}), "fallback local"

print("Ambiente preparado com sucesso.")

Ambiente preparado com sucesso.


## Identificação do grupo

- Integrante 1: João Vitor de Morais Timotio
- Integrante 2: Eduardo Oliveira Silva
- Integrante 3 (se houver): William Cesar Silva de Carvalho
- Integrante 4 (se houver):

## Parte 1 — Explorando a API

Nesta primeira parte, vocês vão consultar 4 endpoints.  
O trabalho do grupo será **alterar apenas a URL** em cada bloco, executar e observar o retorno.


### 1.1 Personagem 10

**Tarefa:** execute o script e observe:
- status code
- nome
- altura
- cor dos olhos

In [8]:
url = "https://swapi.dev/api/people/77/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "San Hill",
    "height": "191",
    "mass": "unknown",
    "hair_color": "none",
    "skin_color": "grey",
    "eye_color": "gold",
    "birth_year": "unknown",
    "gender": "male",
    "homeworld": "https://swapi.dev/api/planets/57/",
    "films": [
        "https://swapi.dev/api/films/5/"
    ],
    "species": [
        "https://swapi.dev/api/species/34/"
    ],
    "vehicles": [],
    "starships": [],
    "created": "2014-12-20T17:58:17.049000Z",
    "edited": "2014-12-20T21:17:50.484000Z",
    "url": "https://swapi.dev/api/people/77/"
}


**Respostas do aluno — Personagem 77**

- Status code: 200
- Nome do personagem: San Hill
- Altura: 191
- Cor dos olhos:gold
- O JSON parece completo? Sim

### 1.2 Personagem 13

**Tarefa:** neste bloco, troque apenas o endpoint se necessário, execute e observe o retorno.

In [13]:
url = "https://swapi.dev/api/people/74/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Jocasta Nu",
    "height": "167",
    "mass": "unknown",
    "hair_color": "white",
    "skin_color": "fair",
    "eye_color": "blue",
    "birth_year": "unknown",
    "gender": "female",
    "homeworld": "https://swapi.dev/api/planets/9/",
    "films": [
        "https://swapi.dev/api/films/5/"
    ],
    "species": [
        "https://swapi.dev/api/species/1/"
    ],
    "vehicles": [],
    "starships": [],
    "created": "2014-12-20T17:32:51.996000Z",
    "edited": "2014-12-20T21:17:50.476000Z",
    "url": "https://swapi.dev/api/people/74/"
}


**Respostas do aluno — Personagem 74**

- Status code: 200
- Nome do personagem:Jocasta Nu
- Altura: 167
- Cor dos olhos: blue
- O JSON parece completo? Sim

### 1.3 Planeta 2

**Tarefa:** execute o bloco e observe:
- nome do planeta
- clima
- terreno
- população

In [14]:
url = "https://swapi.dev/api/planets/2/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Alderaan",
    "rotation_period": "24",
    "orbital_period": "364",
    "diameter": "12500",
    "climate": "temperate",
    "gravity": "1 standard",
    "terrain": "grasslands, mountains",
    "surface_water": "40",
    "population": "2000000000",
    "residents": [
        "https://swapi.dev/api/people/5/",
        "https://swapi.dev/api/people/68/",
        "https://swapi.dev/api/people/81/"
    ],
    "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/6/"
    ],
    "created": "2014-12-10T11:35:48.479000Z",
    "edited": "2014-12-20T20:58:18.420000Z",
    "url": "https://swapi.dev/api/planets/2/"
}


**Respostas do aluno — Planeta 2**

- Status code:200
- Nome do planeta: Alderaan
- Clima: temperate
- Terreno: grasslands, mountains
- População: 2000000000
- O JSON parece completo? Sim

### 1.4 Nave 10

**Tarefa:** execute o bloco e observe:
- nome da nave
- modelo
- fabricante
- velocidade máxima em atmosfera

In [15]:
url = "https://swapi.dev/api/starships/5/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Sentinel-class landing craft",
    "model": "Sentinel-class landing craft",
    "manufacturer": "Sienar Fleet Systems, Cyngus Spaceworks",
    "cost_in_credits": "240000",
    "length": "38",
    "max_atmosphering_speed": "1000",
    "crew": "5",
    "passengers": "75",
    "cargo_capacity": "180000",
    "consumables": "1 month",
    "hyperdrive_rating": "1.0",
    "MGLT": "70",
    "starship_class": "landing craft",
    "pilots": [],
    "films": [
        "https://swapi.dev/api/films/1/"
    ],
    "created": "2014-12-10T15:48:00.586000Z",
    "edited": "2014-12-20T21:23:49.873000Z",
    "url": "https://swapi.dev/api/starships/5/"
}


**Respostas do aluno — Nave 05**

- Status code: 200
- Nome da nave: Sentinel-class landing craft
- Modelo: Sentinel-class landing craft
- Fabricante: Sienar Fleet Systems, Cyngus Spaceworks
- Velocidade máxima em atmosfera: 1000
- O JSON parece completo? Sim

## Parte 2 — Extraindo campos específicos

Agora você vai executar os scripts e **completar as respostas**.

### 2.1 Extraindo campos do personagem

**Tarefa:** execute o bloco e depois responda.

In [16]:
url = "https://swapi.dev/api/people/10/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Nome:", dados["name"])
print("Altura:", dados["height"])
print("Cor dos olhos:", dados["eye_color"])
print("Gênero:", dados["gender"])

Nome: Obi-Wan Kenobi
Altura: 182
Cor dos olhos: blue-gray
Gênero: male


**Perguntas**
1. Qual campo foi usado para mostrar o nome? name
2. Qual campo foi usado para mostrar a altura? height
3. Se você fosse QA, quais desses campos validaria primeiro? O campo name, por causa da relevância da informação

### 2.2 Extraindo campos do planeta

**Tarefa:** execute o bloco e complete as respostas.

In [ ]:
url = "https://swapi.dev/api/planets/2/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Nome do planeta:", dados["name"])
print("Clima:", dados["climate"])
print("Terreno:", dados["terrain"])
print("População:", dados["population"])

Nome do planeta: Alderaan
Clima: temperate
Terreno: grasslands, mountains
População: 2000000000


**Perguntas**
1. O campo população parece numérico? Sim
2. O clima foi retornado corretamente? Sim
3. Se você fosse QA, o que validaria nesse planeta? Tipo de dado de climate e terrain. Validaria se o sistema suporta múltiplos valores

## Parte 3 — Pensando como QA

Agora você vai trabalhar com validações simples.

### 3.1 Validações básicas

**Tarefa:** execute o bloco abaixo e depois escreva o que cada validação verifica.

In [17]:
url = "https://swapi.dev/api/people/10/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Validação 1 - status 200:", response.status_code == 200)
print("Validação 2 - campo 'name' existe:", "name" in dados)
print("Validação 3 - nome não está vazio:", dados["name"] != "")
print("Validação 4 - campo 'height' existe:", "height" in dados)

Validação 1 - status 200: True
Validação 2 - campo 'name' existe: True
Validação 3 - nome não está vazio: True
Validação 4 - campo 'height' existe: True


**Explique com suas palavras**
1. O que a validação 1 verifica? O status
2. O que a validação 2 verifica? Se o campo "name" existe
3. O que a validação 3 verifica? Se o campo "name" está vazio
4. O que a validação 4 verifica? Se o campo "Height" existe

### 3.2 Complete as validações

**Tarefa:** execute o bloco abaixo. Ele já está funcionando. Depois, alterem as expressões se o professor pedir outra nave ou outro endpoint. Em seguida, expliquem por que cada validação faz sentido para QA.


In [22]:
url = "https://swapi.dev/api/starships/10/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Status 200:", response.status_code == 200)

# Os exemplos abaixo já estão corretos e executáveis.
# Se o professor pedir, vocês podem alterar para validar outros campos.
print("Campo 'name' existe:", "name" in dados)
print("Campo 'model' existe:", "model" in dados)
print("Nome não está vazio:", dados["name"] != "")


Status 200: True
Campo 'name' existe: True
Campo 'model' existe: True
Nome não está vazio: True


**Explicando validações**

Status 200: Garante que a comunicação com o servidor foi bem-sucedida antes de processar qualquer dado.

Campos existem (name/model): Valida o contrato da API, assegurando que as chaves esperadas pelo sistema estão presentes no JSON.

Nome não vazio: Certifica que o campo contém informação útil para o usuário, evitando exibir uma interface com dados em branco.

## Parte 4 — Testes automatizados com assert

Agora vocês vão trabalhar sobre scripts já prontos. Os asserts abaixo já executam sem erro. A tarefa do grupo é **entender o que cada linha valida**, ajustar se o professor pedir outro recurso e registrar as respostas.


### 4.1 Assert no personagem

**Tarefa:** execute o bloco. Depois, expliquem o que cada assert valida e, se quiserem, alterem o endpoint para outro personagem.


In [23]:
url = "https://swapi.dev/api/people/13/"
response, origem = buscar_endpoint(url)
dados = response.json()

assert response.status_code == 200, "Erro: status inesperado"
assert "name" in dados, "Erro: campo name não encontrado"
assert dados["name"] != "", "Erro: nome vazio"

print("Todos os testes do personagem passaram!")


Todos os testes do personagem passaram!


**Resposta:**

status_code == 200: Valida se a conexão com o servidor foi bem-sucedida e a API está online.

"name" in dados: Valida se a estrutura do JSON (contrato) contém a chave obrigatória para identificação.

dados["name"] != "": Valida se a informação entregue ao usuário final é preenchida e útil, não apenas nula.

### 4.2 Assert na nave

**Tarefa:** execute o bloco. Depois, descrevam qual campo foi validado em cada assert e proponham mais 1 validação extra em Markdown.


In [ ]:
url = "https://swapi.dev/api/starships/10/"
response, origem = buscar_endpoint(url)
dados = response.json()

assert response.status_code == 200, "Erro: status inesperado"
assert "manufacturer" in dados, "Erro: fabricante não encontrado"
assert dados["model"] != "", "Erro: modelo vazio"

print("Todos os testes da nave passaram!")


Todos os testes da nave passaram!


**Resposta: **

status_code == 200: Valida a conectividade, garantindo que o servidor respondeu com sucesso à requisição.

"manufacturer" in dados: Valida a estrutura, confirmando que a chave do fabricante existe no contrato da API.

dados["model"] != "": Valida a integridade, assegurando que o campo de modelo contém um valor legível e útil.

## Parte 5 — Cenário negativo

Agora você vai observar como a API se comporta quando pedimos um recurso inexistente.

In [ ]:
url = "https://swapi.dev/api/people/9999/"
response, origem = buscar_endpoint(url)

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print("Texto retornado:")
print(response.text)

Origem dos dados: fallback local
Status Code: 404
Texto retornado:
{"detail": "Not found"}


**Perguntas**
1. Qual status code foi retornado? O status code retornado é 404
2. Esse comportamento está correto? Sim, pois o ID 9999 não existe na base de dados.
3. Por que um QA deve testar cenários negativos? Para garantir que o sistema trate erros de forma tranquila, sem expor falhas na segurança.

## Parte 6 — Mini desafio final

Agora você vai alterar apenas o endpoint da função abaixo e gerar um pequeno resumo.

In [24]:
def resumo_recurso(url):
    response, origem = buscar_endpoint(url)
    print("Origem dos dados:", origem)
    print("Status Code:", response.status_code)

    if response.status_code != 200:
        print("Recurso não encontrado.")
        return

    dados = response.json()

    for chave, valor in dados.items():
        print(f"{chave}: {valor}")

# ALTERE O ENDPOINT ABAIXO PARA TESTAR OUTRO RECURSO
resumo_recurso("https://swapi.dev/api/planets/4/")

Origem dos dados: online
Status Code: 200
name: Hoth
rotation_period: 23
orbital_period: 549
diameter: 7200
climate: frozen
gravity: 1.1 standard
terrain: tundra, ice caves, mountain ranges
surface_water: 100
population: unknown
residents: []
films: ['https://swapi.dev/api/films/2/']
created: 2014-12-10T11:39:13.934000Z
edited: 2014-12-20T20:58:18.423000Z
url: https://swapi.dev/api/planets/4/


**Respostas finais**
1. Qual endpoint você testou no desafio? https://swapi.dev/api/planets/4/
2. Qual foi o status code? 200 OK.
3. Qual campo chamou mais a sua atenção? O campo residents veio uma lista vazia.
4. Se você fosse QA, quais 3 validações faria nesse recurso? 1- Tratamento de listas vazias. 2- Tipo de dado em diameter e 3- Consistência de climate

# Entrega da atividade

Entreguem:

1. O notebook preenchido  
2. As respostas em Markdown  
3. Os scripts ajustados  
4. Os asserts analisados e/ou adaptados  
5. Uma conclusão curta do grupo: **quais foram as 3 validações mais importantes?**

**Resposta das 3 validações:** 1- Conectividade (Status Code 200): É o ponto de partida obrigatório. 2- Estrutura do Contrato (Existência de Campos): Consiste em verificar se chaves essenciais (como name, model ou climate) estão presentes no JSON. E 3- Integridade do Conteúdo (Sanidade): Garante que os dados sejam úteis e coerentes.

## Fechamento
Ao final desta atividade, vocês praticaram:

- leitura de resposta JSON
- análise de status code
- validação de campos
- pensamento de QA
- testes simples com assert
- análise de cenário negativo
